### PR TITLE
Gerbera Web UI - Initial Version

### PR DESCRIPTION
**Pull request by elmodaddyb**
Gerbera Web UI - Initial version of new ui #84:

* Add tests with Karma/Jasmine framework
* Initial commit of Gerbera Web - includes Login capability
* Stabilize promise flow for login, implement test specs and e2e with mock API
* Add logout capability, eslint, reformatting
* Add directory and container list, change to BS3, stability improvements
* Update readme with latest test output
* Include first version of tree and items
* Migrate public resources to /web for improved testing
* Add child tree functionality
* Improve usability of menu and tree
* Add thumbnails to items list
* Add highlight item, enter key for login, improve general usability
* Add README for Gerbera Web UI

### CODE REVIEW

**Comment by elmodaddyb** [comment]
Here is a summary of the pull request.

## Gerbera Web UI

The Gerbera Web UI exists under the `/web/gerbera.html` and `/web/js/gerbera*.js` folders alongside the existing user interface implementation.  These components install as part of the existing `make install` process.  The user can access the Gerbera Web UI from a running instance by visiting `http://localhost:49152/gerbera.html`.

#### Single Page Application

The Gerbera Web UI exists under the `/web/gerbera.html` and is the single entry point to the application.

#### Client-side Application Components

The Gerbera UI application components exist in the `/web/js/gerbera*.js` location.  The components are separated similar to the existing implementation.

- App - `gerbera.app.js`
- Auth - `gerbera.auth.js`
- Tree - `gerbera.tree.js` and `jquery.gerbera.tree.js`
- Items - `gerbera.items.js` and `jquery.gerbera.items.js`
- Menu - `gerbera.menu.js`

The only **shared** file is the `/web/js/md5.js`.  All other components are new additions and do not rely on the existing user interface behavior.  

#### Vendor Files

The `/web/vendor` directory contains the third-party libraries for the user interface.  This includes:

- jQuery
- jQuery UI
- jQuery Cookie (plugin)
- Bootstrap

## Test Driven Development

The user interface is developed under the `/gerbera-web` folder that contains a NodeJS project with test-driven suite.  This includes unit tests, end-to-end tests, and a mock server.

**Comment by elmodaddyb** [comment]
I tested the user interface deployment in the following browsers:

- Firefox 55.0.2 (64-bit)
- Google Chrome Version 60.0.3112.113 (Official Build) (64-bit)
- Safari 10.1.2 (macOS)

**Comment by elmodaddyb** [comment]
Here is a report of the test specifications:

## Unit Tests

```
03 09 2017 11:02:45.845:INFO [karma]: Karma v1.7.1 server started at http://0.0.0.0:9876/
03 09 2017 11:02:45.847:INFO [launcher]: Launching browser PhantomJS with unlimited concurrency
03 09 2017 11:02:45.851:INFO [launcher]: Starting browser PhantomJS
03 09 2017 11:02:46.052:INFO [PhantomJS 2.1.1 (Linux 0.0.0)]: Connected on socket jO9UEcRg8bAj7zlYAAAA with id 24531835
  Gerbera UI App
    initialize()
      ✔ retrieves the configuration from the server using AJAX
      ✔ stores the TYPE and SID cookies to the document
      ✔ defaults the TYPE to `db` when none is set
  Gerbera Auth
    checkSID()
      ✔ calls the server for a new SID and sets the cookie for SID
      ✔ calls GERBERA.App.error when the server call fails
    SID()
      ✔ retrieves the session ID from the cookie
    authenticate()
      ✔ calls the server to get a token and then logs the user into the system
    logout()
      ✔ calls the server to logout the session and expires the session ID cookie
  Gerbera Items
    initialize()
      ✔ loads the # of view items based on default(25) when app config is undefined
      ✔ loads the # of view items based on application config
    treeItemSelected()
      ✔ updates the breadcrumb based on the selected item
    transformItems()
      ✔ converts a gerbera list of items to the widget accepted list
    loadItems()
      ✔ does not load items if response is failure
      ✔ loads the response as items in the datagrid
  Gerbera Menu
    initialize()
      ✔ binds all menu items with the click event
      ✔ on click of Database calls to load the containers
03 09 2017 11:02:46.333:WARN [web-server]: 404: /content/interface?req_type=containers&parent_id=0&select_it=0
      ✔ on click of menu, clears items
LOG: 'failure: [object Object]'
      ✔ on click of home menu, clears tree and items
  Gerbera Tree
    initialize()
      ✔ loads the tree view
    loadTree()
      ✔ renders the tree with parent and children
      ✔ does not render the tree when response is not successful
      ✔ clears the breadcrumb and loads the first item on load of a new tree
    transformContainers
      ✔ transforms containers response suitable for treeview load
      ✔ only creates parent when boolean is passed
    selectType()
      ✔ calls the server for containers when type is db
      ✔ calls the server for containers when type is db 
      ✔ on failure calls the app error handler
    onItemSelected()
      ✔ updates the breadcrumb based on the click of a item
      ✔ passes gerbera data to the items plugin
      ✔ changes the color of the selected row
    onItemExpanded()
      ✔ calls for child items based on the response
  The jQuery Datagrid
    ✔ creates a table based on the dataset provided
  The jQuery Tree
    ✔ creates a list the same size as the data
    ✔ items are loaded in folder closed state
    ✔ binds the item title text to a onSelection method
    ✔ binds the folder icon to a onExpand method
    ✔ appends badges to the item
    ✔ recursively appends children based on data lineage
    appendItems()
      ✔ given a parent item append children as list
    clear()
      ✔ removes all contents of the tree
    closed()
      ✔ determines whether a node is closed
    collapse()
      ✔ given a parent item append children as list
    select()
      ✔ selects an item by changing the style

Finished in 0.29 secs / 0.29 secs @ 11:02:46 GMT-0400 (EDT)

SUMMARY:
✔ 43 tests completed
```

## End-to-End Tests

```
    
    > gerbera-web@0.0.1 start /development/gerbera/gerbera-web
    > node index.js
    
    
    > gerbera-web@0.0.1 e2e /development/gerbera/gerbera-web
    > mocha "test/e2e"
    
    Mock webserver listenting on : 3000
    
    
      The home page
        ✓ without logging in the database and filetype menus are disabled (71ms)
        ✓ enables the database and filetype menus upon login (397ms)
        ✓ disables the database and filetype menus upon logout (314ms)
        ✓ loads the parent database container list when clicking Database (446ms)
        ✓ loads the parent filesystem container list when clicking Filesystem (116ms)
        ✓ clears the tree when Home menu is clicked (116ms)
        ✓ loads the related items when tree item is clicked (182ms)
        ✓ loads child tree items when tree item is expanded (713ms)
    
      The login page
        ✓ hides the login form when no authentication is required (48ms)
        ✓ shows the login form when no session exists yet and accounts is required (158ms)
        ✓ requires user name and password to submit the login form (209ms)
        ✓ when successful login show logout button, and show form on logout (475ms)
    
    
      12 passing (6s)
```

**Comment by v00d00** [comment]
Christmas has come early 😁 🎄🎄

I will look at this tomorrow.

**Comment by v00d00** [comment]
Overall, the main UI with the tree and everything looks great.

Some thoughts:

1. At login there is a load of stuff that isnt really useful until you log in, can we tidy it up to make it more obvious?
2. When trying to login in the failure message is at the bottom of the page.
3. Does it run without any users defined?
4. To expand an item in the tree you cant click the label, only the folder icon.

**Comment by elmodaddyb** [comment]
Thanks for checking the ui system.  I will fix as you mentioned.

1. Make failure message top aligned
2. Remove howto info and hide/disable menus until logged in.
3. Improve expand of tree using label + icon.

I will double check the automatic login with `accounts` are set to `no`.  I think that is working as existing system.

**Comment by v00d00** [comment]
All needs is the "add file" and "add autoscan" parts and then we can make it the main UI!

File: web/gerbera.html [hunk 1]
```diff
@@ -23,9 +23,9 @@
         </div>
         <div class="collapse navbar-collapse">
             <ul class="nav navbar-nav navbar-left">
-                <li><a id="nav-home" class="navbar-link" data-gerbera-menu-cmd="HOME">Home <span class="sr-only">(current)</span></a></li>
-                <li><a id="nav-db" class="navbar-link disabled" data-gerbera-menu-cmd="SELECT_TYPE" data-gerbera-type="db">Database</a></li>
-                <li><a id="nav-fs" class="navbar-link disabled" data-gerbera-menu-cmd="SELECT_TYPE" data-gerbera-type="fs">File System</a></li>
+                <li><a id="nav-home" class="btn navbar-link" data-gerbera-menu-cmd="HOME">Home <span class="sr-only">(current)</span></a></li>
```
Comment (elmodaddyb) [inline web/gerbera.html]: Adding the `btn` class makes the `disabled` class work properly.  Previously, `disabled` was not getting applied.

File: web/gerbera.html [hunk 2]
```diff
@@ -57,6 +57,15 @@
   })
 </script>
 
+<div class="container-fluid">
```
Comment (elmodaddyb) [inline web/gerbera.html]: Moved `warning` to top, right below menu bar.  Added `close` button on right side.

File: web/gerbera.html [hunk 3]
```diff
@@ -68,34 +77,15 @@
     <div class="row">
         <div class="col-md-5">
             <div id="tree">
-                <div class="jumbotron">The Gerbera Tree</div>
```
Comment (elmodaddyb) [inline web/gerbera.html]: This info should go to **gerbera.io** in some other capacity.  Removing....

File: web/js/gerbera.menu.js [hunk 4]
```diff
@@ -11,6 +11,7 @@ GERBERA.Menu = (function () {
     if (GERBERA.Auth.isLoggedIn()) {
       allLinks.click(GERBERA.Menu.click)
       allLinks.removeClass('disabled')
+      $('#nav-home').click()
```
Comment (elmodaddyb) [inline web/js/gerbera.menu.js]: By default, once logged in, the user goes to the `Home` menu option.

File: web/js/gerbera.tree.js [hunk 5]
```diff
@@ -76,8 +77,9 @@ GERBERA.Tree = (function () {
       .fail(GERBERA.App.error)
   }
 
-  var loadTree = function (response) {
+  var loadTree = function (response, config) {
     var tree = $('#tree')
+    config = $.extend({}, treeViewConfig, config)
```
Comment (elmodaddyb) [inline web/js/gerbera.tree.js]: Added extend here to make `treeViewConfig` more testable after adding both `onExpand` and `onSelection` to the same click.

File: web/js/gerbera.tree.js [hunk 6]
```diff
@@ -158,9 +160,9 @@ GERBERA.Tree = (function () {
   var destroy = function () {
     currentTree = []
     currentType = 'db'
-
-    $('#tree').tree('destroy')
-
+    if($('#tree').hasClass('grb-tree')) {
+      $('#tree').tree('destroy')
```
Comment (elmodaddyb) [inline web/js/gerbera.tree.js]: Squashed bug found during testing.

**Comment by elmodaddyb** [comment]
@v00d00  🥇 for the Gerbera organization!!

I cleaned up the UI as discussed via b7041615124a09405c3907d7e1ac8f4276978ef1.  

I tested the system automatically logs the user in when `<accounts enabled="no"/>` and properly hides all login/logout elements.

The highlighting of the selected tree item bleeds a bit due to default bootstrap stuff....but its tolerable for the initial version.  I am presently working on a Sketch + Zeplin view of Gerbera to allow for clear definition of the CSS.  This should help solve some of the CSS.

**Comment by v00d00** [comment]
Thanks!

**Pull request merged**